tuning/trainercontroller/callback.py

```diff
@@ -0,0 +1,353 @@
+        """
+        for handler in handlers:
+            self.metric_handlers[handler.__name__]=handler
+
+    def register_operation_handlers(self, operation_handlers: List[Operation]):
+        """Registers the operation handlers
+
+        Args:
+            operation_handlers: List[Operation]. List of operation handlers.
+        """
+        for operation_handler in operation_handlers:
+            self.operation_handlers[operation_handler.__name__]=operation_handler
+
+    def _compute_metrics(self, event_name: str, **kwargs):
+        """Invokes the compute() for all the metrics registered for a given event.
+
+        Args:
+            event_name: str. Event name.
+        """
+        if event_name in self.metrics_on_event:
+            for m in self.metrics_on_event[event_name]:
+                self.metrics[m.get_name()] = m.compute(event_name=event_name, **kwargs)
+
+    def _take_control_actions(self, event_name: str, **kwargs):
+        """Invokes the act() method for all the operations registered for a given event. Note here that the eval() is invoked with `__builtins__` set to None. This is a precaution to restric the scope of eval(), to only the fields produced by the metrics.
+
+        Args:
+            event_name: str. Event name.
+            kwargs: List of arguments (key, value)-pairs.
+        """
+        if event_name in self.control_actions_on_event:
+            for control_action in self.control_actions_on_event[event_name]:
+                rule_succeeded = False
+                try:
+                    rule_succeeded = eval(control_action.rule, {'__builtins__': None}, self.metrics)
+                except TypeError:
+                    raise TypeError("Rule failed due to incorrect type usage")
+                except ValueError:
+                    raise ValueError("Rule failed due to use of disallowed packages")
+                except NameError:
+                    raise NameError("Rule failed due to use of disallowed variables")
+                if rule_succeeded:
+                    for operation_action in control_action.operation_actions:
+                        logger.info(f"Taking {operation_action.action} \
+                                    action in {control_action.name}")
+                        operation_action.instance.act(action=operation_action.action, event_name=event_name, **kwargs)
+
+    def _actions_on_event(self, event_name: str, **kwargs):
+        """Invokes all functions associated with an event.
+
+        Args:
+            event_name: str. Event name.
+            kwargs: List of arguments (key, value)-pairs.
+        """
+        self._compute_metrics(event_name, **kwargs)
+        self._take_control_actions(event_name, **kwargs)
+
+    def _validate_rule(self, rule):
+        """Validates the rule to check if there are any import attempts 
+
+        Returns:
+            bool
+        """
+        return re.search(r'__', rule) is None
```

Comment:
I think I don't understand this correctly - can you explain how this regex is checking import attempts (e.g., with an example of something that is caught by this that is potentially malicious)?

Reply:
@alex-jw-brooks From what we understand on the use of `eval()`, sanitizing the input is one of the ways to mitigate the risk of using `eval()`.

To that end, we have added this additional layer of check to prevent the use of variables like `__code__`, `__subclasses__` etc which could potentially leak information.

Reply:
Ahhh I understand now, I forgot about the `__import__` function hook and thought that this meant the `import` statement. Thanks, that helps, makes sense